man/repl.man
.TH REPL %manext1% 2014-12-15 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
repl \- reply to an nmh message
.SH SYNOPSIS
.HP 5
.na
.B repl
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msg ]
.RB [ \-annotate " | " \-noannotate ]
.RB [ \-group " | " \-nogroup ]
.RB [ \-cc
all/to/cc/me]
.RB [ \-nocc
all/to/cc/me]
.RB [ \-query " | " \-noquery ]
.RB [ \-form
.IR formfile ]
.RB [ \-format " | " \-noformat ]
.RB [ \-filter
.IR filterfile ]
.RB [ \-inplace " | " \-noinplace ]
.RB [ \-mime " | " \-nomime ]
.RB [ \-fcc
.IR +folder ]
.RB [ \-width
.IR columns ]
.RB [ \-draftfolder
.IR +folder ]
.RB [ \-draftmessage
.IR msg ]
.RB [ \-nodraftfolder ]
.RB [ \-editor
.IR editor ]
.RB [ \-noedit ]
.RB [ \-convertargs
.IR "type argstring" ]
.RB [ \-whatnowproc
.IR program ]
.RB [ \-nowhatnowproc ]
.RB [ \-atfile ]
.RB [ \-noatfile ]
.RB [ \-fmtproc
.IR program ]
.RB [ \-nofmtproc ]
.RB [ \-build ]
.RB [ \-file
.IR msgfile ]
.ad
.SH DESCRIPTION
.B repl
may be used to reply to a message.
.PP
In its simplest form (with no arguments),
.B repl
will set up a message-form skeleton in reply to the current message
in the current folder, and invoke the
.B whatnow
shell.
.PP
.B repl
uses a reply template to construct the draft of the reply.
A reply template is simply an
.B mhl
format file (see
.IR mh\-format (5)
for details).
.PP
If the switch
.B \-nogroup
is given (it is on by default), then
.B repl
will use the standard forms file \*(lqreplcomps\*(rq.  This will construct
a draft message that is intended to be sent only to the author of the
message to which you are replying.  If a file named \*(lqreplcomps\*(rq
exists in the user's
.B nmh
directory, it will be used instead of this default forms file.
.PP
The default reply template \*(lqreplcomps\*(rq will direct
.B repl
to construct the reply message draft as follows:
.PP
.RS 5
.nf
To: <Mail\-Reply\-To> or <Reply\-To> or <From>
cc: <To> and <cc> and <personal address>
Fcc: {fcc switch} or +outbox
Subject: Re: <Subject>
In\-Reply\-To: <Message\-Id>
References: <Message\-Id>
Comments: In\-Reply\-To <From> or <apparently from> or <Sender>
.ti +3
message dated <date>
--------
.fi
.RE
.PP
where field names enclosed in angle brackets (<\ >) indicate the
contents of the named field from the message to which the reply is
being made.
.PP
By default, the \*(lqcc:\*(rq field is empty.  You may selectively add
addresses to this default with the
.B \-cc
.I type
switch.  This switch takes an argument
.RI ( all / to / cc / me )
which specifies who gets added to the default
\*(lqcc:\*(rq list of the reply.  You may give this switch multiple times (with
different arguments) if you wish to add multiple types of address.
.PP
If the switch
.B \-group
is given, then
.B repl
will use the standard forms file \*(lqreplgroupcomps\*(rq.
This will construct a draft message that is intended as a group or
followup reply.  If a file named \*(lqreplgroupcomps\*(rq exists
in the user's nmh directory, it will be used instead of this default
forms file, unless you specify another forms file on the command line
or in your profile.
.PP
The default group reply template \*(lqreplgroupcomps\*(rq will direct
.B repl
to construct the reply message draft as follows:
.PP
.RS 5
.nf
To: <Mail\-Followup\-To>
Subject: Re: <Subject>
In\-Reply\-To: Message from <From> of <Date>.
.ti +\w'In\-Reply\-To: 'u
<Message\-Id>
--------
.fi
.RE
.PP
or if the field <Mail\-Followup\-To> is not available:
.PP
.RS 5
.nf
To: <Mail\-Reply\-To> or <Reply\-To> or <From>
cc: <To> and <cc> and <personal address>
Subject: Re: <Subject>
In\-Reply\-To: Message from <From> of <Date>.
.ti +\w'In\-Reply\-To: 'u
<Message\-Id>
--------
.fi
.RE
.PP
By default, the \*(lqcc:\*(rq contains all the addresses shown.  You may
selectively remove addresses from this default with the
.B \-nocc
.I type
switch.
This switch takes an argument (
.IR all / to / cc / me )
which specifies who gets removed from the default \*(lqcc:\*(rq list of
the reply.  You may give this switch multiple times (with different arguments)
if you wish to remove multiple types of address.
.PP
In any case, you may specify an alternative forms file with the switch
.B \-form
.IR formfile .
.PP
The
.B \-query
switch modifies the action of
.B \-nocc
.I type
switch by interactively asking you if each address that normally would be
placed in the \*(lqTo:\*(rq and \*(lqcc:\*(rq list should actually be sent a copy.
This is useful for special-purpose replies.  Note that the position of
the
.B \-cc
and
.B \-nocc
switches, like all other switches which take a positive and negative form,
is important.
.PP
Lines beginning with the fields \*(lqTo:\*(rq, \*(lqcc:\*(rq, and
\*(rqBcc:\*(rq will be standardized and have duplicate addresses removed.
In addition, the
.B \-width
.I columns
switch will guide
.BR repl 's
formatting of these fields.
.PP
If the draft already exists,
.B repl
will ask you as to the disposition of the draft.  A reply of
.B quit
will abort
.BR repl ,
leaving the
draft intact;
.B replace
will replace the existing draft with a blank skeleton; and
.B list
will display the draft.
.PP
See
.IR comp (1)
for a description of the
.B \-editor
and
.B \-noedit
switches.  Note that while in the editor, with
.B \-atfile
and if the current directory is writable, the message being replied
to is available through a link named \*(lq@\*(rq (assuming the default
.IR whatnowproc ).
In addition, the actual pathname of the message is stored in the
environment variable
.BR $editalt ,
and the pathname of the folder containing the message is stored in
the environment variable
.BR $mhfolder .
The creation of the \*(lq@\*(rq file is controlled via the
.B \-atfile
and
.B \-noatfile
options.
.PP
The
.B \-convertargs
switch directs
.B repl
to pass the arguments for
.I type
to
.BR mhbuild .
Both arguments are required;
.I type
must be non-empty while
.I argstring
can be empty, e.g., '' in a shell command line.  The
.B \-convertargs
switch can be used multiple times.
See the Convert Interface section of
.IR mhbuild (1)
for a description of the convert mechanism, and
.I %docdir%/contrib/replaliases
for examples of its use.
.PP
Although
.B repl
uses a forms file to direct it how to construct the beginning of the
draft, it uses a message filter file to direct it as to how the message
to which you are replying should be filtered (re-formatted) in the body
of the draft.  The filter file for
.B repl
should be a standard form file for
.BR mhl ,
as
.B repl
will invoke
.B mhl
to format the message to which you are replying.
.PP
The switches
.BR \-noformat ,
.BR \-format ,
and
.B \-filter
.I filterfile
specify which message filter file to use.
.PP
If the switch
.B \-noformat
is given (it is the default) and the
.B \-filter
switch is not used, then the message to which you are replying is not
included in the body of the draft.
.PP
If the switch
.B \-format
is given, then a default message filter file is used.
This default message filter should be adequate for most users.
This default filter
.RI \*(lq mhl.reply \*(rq
is:
.PP
.RS 5
.nf
%mhl_reply%
.fi
.RE
.PP
which outputs each line of the body of the message prefaced with the
\*(lq>\*(rq character and a space.
.PP
If a file named
.RI \*(lq mhl.reply \*(rq
exists in the user's
.B nmh
directory,
it will be used instead of this form.  You may specify an alternate
message filter file with the switch
.B \-filter
.IR filterfile .
.PP
Other reply filters are commonly used, such as:
.PP
.RS 5
.nf
:
body:nocomponent,compwidth=9,offset=9
.fi
.RE
.PP
which says to output a blank line, and then the body of the message
being replied to, indented by one tab stop.  Another popular format
is:
.PP
.RS 5
.nf
message-id:nocomponent,\|nonewline,\\
formatfield=\*(lqIn message %{text},\ \*(rq
from:nocomponent,\|formatfield=\*(lq%(decode(friendly{text})) writes:\*(rq
body:component=\*(lq>\*(rq,\|overflowtext=\*(lq>\*(rq,\|overflowoffset=0
.fi
.RE
.PP
This message filter file cites the Message-ID and author of the message
being replied to, and then outputs each line of the body prefaced with
the \*(lq>\*(rq character.
.PP
You can also use an external format program to format the message body.
The format program is specified by the
.I formatproc
profile entry, and is enabled by the \*(lqformat\*(rq flag.  A message
filter using an external format program would look like this:
.PP
.RS 5
.nf
body:component=\*(lq>\*(rq,\|nowrap,\|format
.fi
.RE
.PP
See the
.IR mhl (1)
documentation for more information.
The format program can be changed by the
.B \-fmtproc
.I program
and
.B \-nofmtproc
switches.
.PP
To use the MIME rules for encapsulation, specify the
.B \-mime
switch.
This directs
.B repl
to generate an
.B mhbuild
composition file.
Note that
.B nmh
will not invoke
.B mhbuild
automatically; you must specifically give the command
.PP
.RS 5
What now? mime
.RE
.PP
prior to sending the draft.
.PP
If the
.B \-annotate
switch is given, the message being replied to will be annotated with
the lines
.PP
.RS 5
Replied:\ date
Replied:\ addrs
.RE
.PP
where the address list contains one line for each addressee.
The annotation will be done only if the message is sent directly from
.BR repl .
If the message is not sent immediately from
.BR repl ,
.RB \*(lq "comp\ \-use" \*(rq
may be used to re-edit and send the constructed message, but the
annotations won't take place.
Normally annotations are done inplace in order to preserve any links
to the message.
You may use the
.B \-noinplace
switch to change this.
.PP
Although the default template specifies that a copy of the reply will be
put in the folder 'outbox', if the
.B \-fcc
.I +folder
switch is given it will override the default value.
More than one folder, each preceded by
.B \-fcc
can be named.
.PP
In addition to the standard
.IR mh\-format (5)
escapes,
.B repl
also recognizes the following additional
.I component
escape:
.PP
.RS 5
.nf
.ta \w'Escape  'u +\w'Returns  'u
.I "Escape	Returns	Description"
fcc	string	Any folders specified with `\-fcc\ folder'
.fi
.RE
.PP
To avoid reiteration,
.B repl
strips any leading `Re: ' strings from the
.I subject
component.
.PP
The
.B \-draftfolder
.I +folder
and
.B \-draftmessage
.I msg
switches invoke the
.B nmh
draft folder facility.  This is an advanced (and highly useful) feature.
Consult
.IR mh-draft (5)
for more information.
.PP
Upon exiting from the editor,
.B repl
will invoke the
.B whatnow
program.  See
.IR whatnow (1)
for a discussion of available
options.  The invocation of this program can be inhibited by using the
.B \-nowhatnowproc
switch.  (In fact, it is the
.B whatnow
program which starts the initial edit.  Hence,
.B \-nowhatnowproc
will prevent any edit from occurring.)
.PP
The
.B \-build
switch is intended to be used by the Emacs mh-e interface
to
.BR nmh .
It implies
.BR \-nowhatnowproc .
It causes a file
.I <mh-dir>/reply
to be created, containing the draft message that would normally be presented
to the user for editing.
No mail is actually sent.
.PP
The
.B \-file
.I msgfile
switch specifies the message to be replied to as an exact filename rather
than as an
.B nmh
folder and message number.
The same caveats apply to this option as to the
.B \-build
switch.
.SH FILES
.B repl
looks for all format, filter and template files in multiple locations:
absolute pathnames are accessed directly, tilde expansion is done on
usernames, and files are searched for in the user's
.I Mail
directory as specified in their profile.  If not found there, the directory
.RI \*(lq %nmhetcdir% \*(rq
is checked.
.PP
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^%nmhetcdir%/replcomps~^The standard reply template
^or <mh-dir>/replcomps~^Rather than the standard template
^%nmhetcdir%/replgroupcomps~^The standard `reply -group' template
^or <mh-dir>/replgroupcomps~^Rather than the standard template
^%nmhetcdir%/mhl.reply~^The standard message filter
^or <mh-dir>/mhl.reply~^Rather than the standard filter
^$HOME/.mh_profile~^The user profile
^<mh-dir>/draft~^The draft file
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Alternate\-Mailboxes:~^To determine the user's mailboxes
^Current\-Folder:~^To find the default current folder
^Draft\-Folder:~^To find the default draft-folder
^Editor:~^To override the default editor
^Msg\-Protect:~^To set mode when creating a new message (draft)
^fileproc:~^Program to refile the message
^mhlproc:~^Program to filter message being replied-to
^whatnowproc:~^Program to ask the \*(lqWhat now?\*(rq questions
.fi
.SH "SEE ALSO"
.IR comp (1),
.IR forw (1),
.IR mh\-format (5),
.IR mhbuild (1),
.IR send (1),
.IR whatnow (1)
.PP
.I %docdir%/contrib/replaliases
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder"
.RB ` msg "' defaults to cur"
.RB ` \-nogroup '
.RB ` "\-nocc\ all" "' with `\-nogroup', `\-cc\ all' with `\-group'"
.RB ` \-noannotate '
.RB ` \-nodraftfolder '
.RB ` \-noformat '
.RB ` \-inplace '
.RB ` \-nomime '
.RB ` \-noquery '
.RB ` \-noatfile '
.RB ` "\-width\ 72" '
.fi
.SH CONTEXT
If a folder is given, it will become the current folder.  The message
replied to will become the current message.
.SH BUGS
If any addresses occur in the reply template, addresses in the template
that do not contain hosts are defaulted incorrectly.  Instead of using
the localhost for the default,
.B repl
uses the sender's host.
Moral of the story: if you're going to include addresses in a reply
template, include the host portion of the address.
.PP
The
.B \-width
.I columns
switch is only used to do address-folding; other headers are not line-wrapped.
.PP
If
.I whatnowproc
is
.BR whatnow ,
then
.B repl
uses a built-in
.BR whatnow ,
it does not actually run the
.B whatnow
program.
Hence, if you define your own
.IR whatnowproc ,
don't call it
.B whatnow
since
.B repl
won't run it.


NEWS
#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.
- mhfixmsg now adds a Content-Transfer-Encoding header at the message level,
  if needed after decoding text parts.

uip/repl.c
/* repl.c -- reply to a message
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/mime.h>
#include <h/utils.h>
#include "sbr/m_maildir.h"

#define REPL_SWITCHES \
    X("group", 0, GROUPSW) \
    X("nogroup", 0, NGROUPSW) \
    X("annotate", 0, ANNOSW) \
    X("noannotate", 0, NANNOSW) \
    X("cc all|to|cc|me", 0, CCSW) \
    X("nocc all|to|cc|me", 0, NCCSW) \
    X("draftfolder +folder", 0, DFOLDSW) \
    X("draftmessage msg", 0, DMSGSW) \
    X("nodraftfolder", 0, NDFLDSW) \
    X("editor editor", 0, EDITRSW) \
    X("noedit", 0, NEDITSW) \
    X("convertargs type argstring", 0, CONVERTARGSW) \
    X("fcc folder", 0, FCCSW) \
    X("filter filterfile", 0, FILTSW) \
    X("form formfile", 0, FORMSW) \
    X("format", 5, FRMTSW) \
    X("noformat", 7, NFRMTSW) \
    X("inplace", 0, INPLSW) \
    X("noinplace", 0, NINPLSW) \
    X("mime", 0, MIMESW) \
    X("nomime", 0, NMIMESW) \
    X("query", 0, QURYSW) \
    X("noquery", 0, NQURYSW) \
    X("whatnowproc program", 0, WHATSW) \
    X("nowhatnowproc", 0, NWHATSW) \
    X("width columns", 0, WIDTHSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("file file", 4, FILESW) \
    X("build", 5, BILDSW) /* interface from mhe */ \
    X("atfile", 0, ATFILESW) \
    X("noatfile", 0, NOATFILESW) \
    X("fmtproc program", 0, FMTPROCSW) \
    X("nofmtproc", 0, NFMTPROCSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(REPL);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(REPL, switches);
#undef X

#define CC_SWITCHES \
    X("to", 0, CTOSW) \
    X("cc", 0, CCCSW) \
    X("me", 0, CMESW) \
    X("all", 0, CALSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(CC);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(CC, ccswitches);
#undef X

#define DISPO_SWITCHES \
    X("quit", 0, NOSW) \
    X("replace", 0, YESW) \
    X("list", 0, LISTDSW) \
    X("refile +folder", 0, REFILSW) \
    X("new", 0, NEWSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(DISPO);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(DISPO, aqrnl);
#undef X

static struct swit aqrl[] = {
    { "quit", 0, NOSW },
    { "replace", 0, YESW },
    { "list", 0, LISTDSW },
    { "refile +folder", 0, REFILSW },
    { NULL, 0, 0 }
};

short ccto = -1;		/* global for replsbr */
short cccc = -1;
short ccme = -1;
short querysw = 0;

static short outputlinelen = OUTPUTLINELEN;
static short groupreply = 0;		/* Is this a group reply?        */

static int mime = 0;			/* include original as MIME part */
static char *form   = NULL;		/* form (components) file        */
static char *filter = NULL;		/* message filter file           */
static char *fcc    = NULL;		/* folders to add to Fcc: header */


/*
 * prototypes
 */
static void docc (char *, int);
static void add_convert_header (const char *, char *, char *, char *);


int
main (int argc, char **argv)
{
    int	i, isdf = 0;
    int anot = 0, inplace = 1;
    int nedit = 0, nwhat = 0;
    int atfile = 0;
    int fmtproc = -1;
    char *cp, *cwd, *dp, *maildir, *file = NULL;
    char *folder = NULL, *msg = NULL, *dfolder = NULL;
    char *dmsg = NULL, *ed = NULL, drft[BUFSIZ], buf[BUFSIZ];
    char **argp, **arguments;
    svector_t convert_types = svector_create (10);
    svector_t convert_args = svector_create (10);
    size_t n;
    struct msgs *mp = NULL;
    struct stat st;
    FILE *in;

    int buildsw = 0;

    if (nmh_init(argv[0], 1)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s: [+folder] [msg] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case GROUPSW:
		    groupreply++;
		    continue;
		case NGROUPSW:
		    groupreply = 0;
		    continue;

		case ANNOSW: 
		    anot++;
		    continue;
		case NANNOSW: 
		    anot = 0;
		    continue;

		case CCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    docc (cp, 1);
		    continue;
		case NCCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    docc (cp, 0);
		    continue;

		case EDITRSW: 
		    if (!(ed = *argp++) || *ed == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nedit = 0;
		    continue;
		case NEDITSW:
		    nedit++;
		    continue;
		    
		case CONVERTARGSW: {
                    char *type;
                    size_t i;

		    if (!(type = *argp++)) {
			adios (NULL, "missing type argument to %s", argp[-2]);
                    }
		    if (!(cp = *argp++)) {
			adios (NULL, "missing argstring argument to %s",
			       argp[-3]);
                    }

                    for (i = 0; i < svector_size (convert_types); ++i) {
                        if (! strcmp (svector_at (convert_types, i), type)) {
                            /* Already saw this type, so just update
                               its args. */
                            svector_strs (convert_args)[i] = cp;
                            break;
                        }
                    }

                    if (i == svector_size (convert_types)) {
                        svector_push_back (convert_types, type);
                        svector_push_back (convert_args, cp);
                    }
		    continue;
                }

		case WHATSW: 
		    if (!(whatnowproc = *argp++) || *whatnowproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nwhat = 0;
		    continue;
		case BILDSW: 
		    buildsw++;
		    /* FALLTHRU */
		case NWHATSW: 
		    nwhat++;
		    continue;

		case FCCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    dp = NULL;
		    if (*cp == '@')
			cp = dp = path (cp + 1, TSUBCWF);
		    if (fcc)
			fcc = add (", ", fcc);
		    fcc = add (cp, fcc);
                    free(dp);
		    continue;

		case FILESW: 
		    if (file)
			adios (NULL, "only one file at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    file = path (cp, TFILE);
		    continue;
		case FILTSW:
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    filter = getcpy (etcpath (cp));
		    mime = 0;
		    continue;
		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case FRMTSW: 
		    filter = getcpy (etcpath (mhlreply));
		    mime = 0;
		    continue;
		case NFRMTSW: 
		    filter = NULL;
		    continue;

		case INPLSW: 
		    inplace++;
		    continue;
		case NINPLSW: 
		    inplace = 0;
		    continue;

		case MIMESW:
		    mime++;
		    filter = NULL;
		    continue;
		case NMIMESW:
		    mime = 0;
		    continue;

		case QURYSW: 
		    querysw++;
		    continue;
		case NQURYSW: 
		    querysw = 0;
		    continue;

		case WIDTHSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if ((outputlinelen = atoi (cp)) < 10)
			adios (NULL, "impossible width %d", outputlinelen);
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			adios (NULL, "only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
				    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW:
		    if (dmsg)
			adios (NULL, "only one draft message at a time!");
		    if (!(dmsg = *argp++) || *dmsg == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;

		case ATFILESW:
		    atfile++;
		    continue;
		case NOATFILESW:
		    atfile = 0;
		    continue;

		case FMTPROCSW:
		    if (!(formatproc = *argp++) || *formatproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    fmtproc = 1;
		    continue;
		case NFMTPROCSW:
		    fmtproc = 0;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
            folder = pluspath (cp);
	} else {
	    if (msg)
		adios (NULL, "only one message at a time!");
            msg = cp;
	}
    }

    if (ccto == -1) 
	ccto = groupreply;
    if (cccc == -1)
	cccc = groupreply;
    if (ccme == -1)
	ccme = groupreply;

    cwd = mh_xstrdup(pwd ());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (file && (msg || folder))
	adios (NULL, "can't mix files and folders/msgs");

try_it_again:

    strncpy (drft, buildsw ? m_maildir ("reply")
			  : m_draft (dfolder, NULL, NOUSE, &isdf), sizeof(drft));

    /* Check if a draft exists */
    if (!buildsw && stat (drft, &st) != NOTOK) {
	printf ("Draft \"%s\" exists (%ld bytes).", drft, (long) st.st_size);
	for (i = LISTDSW; i != YESW;) {
	    if (!(argp = read_switch_multiword ("\nDisposition? ",
						isdf ? aqrnl : aqrl)))
		done (1);
	    switch (i = smatch (*argp, isdf ? aqrnl : aqrl)) {
		case NOSW: 
		    done (0);
		case NEWSW: 
		    dmsg = NULL;
		    goto try_it_again;
		case YESW: 
		    break;
		case LISTDSW: 
		    showfile (++argp, drft);
		    break;
		case REFILSW: 
		    if (refile (++argp, drft) == 0)
			i = YESW;
		    break;
		default: 
		    inform("say what?");
		    break;
	    }
	}
    }

    if (file) {
	/*
	 * We are replying to a file.
	 */
	anot = 0;	/* we don't want to annotate a file */
    } else {
	/*
	 * We are replying to a message.
	 */
	if (!msg)
	    msg = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder, 1)))
	    adios (NULL, "unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    adios (NULL, "no messages in %s", folder);

	/* parse the message range/sequence/name and set SELECTED */
	if (!m_convert (mp, msg))
	    done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	if (mp->numsel > 1)
	    adios (NULL, "only one message at a time!");

	context_replace (pfolder, folder);	/* update current folder   */
	seq_setcur (mp, mp->lowsel);	/* update current message  */
	seq_save (mp);			/* synchronize sequences   */
	context_save ();			/* save the context file   */
    }

    msg = file ? file : mh_xstrdup(m_name (mp->lowsel));

    if ((in = fopen (msg, "r")) == NULL)
	adios (msg, "unable to open");

    /* find form (components) file */
    if (!form) {
	if (groupreply)
	    form = etcpath (replgroupcomps);
	else
	    form = etcpath (replcomps);
    }

    replout (in, msg, drft, mp, outputlinelen, mime, form, filter,
    	     fcc, fmtproc);
    fclose (in);

    {
        char *filename = file ? file : concat (mp->foldpath, "/", msg, NULL);

        for (n = 0; n < svector_size (convert_types); ++n) {
            add_convert_header (svector_at (convert_types, n),
                                svector_at (convert_args, n),
                                filename, drft);
        }
        if (! file) {
            free (filename);
        }
    }

    if (nwhat)
	done (0);
    what_now (ed, nedit, NOUSE, drft, msg, 0, mp, anot ? "Replied" : NULL,
	    inplace, cwd, atfile);

    svector_free (convert_args);
    svector_free (convert_types);

    done (1);
    return 1;
}

void
docc (char *cp, int ccflag)
{
    switch (smatch (cp, ccswitches)) {
	case AMBIGSW: 
	    ambigsw (cp, ccswitches);
	    done (1);
	case UNKWNSW: 
	    adios (NULL, "-%scc %s unknown", ccflag ? "" : "no", cp);

	case CTOSW: 
	    ccto = ccflag;
	    break;

	case CCCSW: 
	    cccc = ccflag;
	    break;

	case CMESW: 
	    ccme = ccflag;
	    break;

	case CALSW: 
	    ccto = cccc = ccme = ccflag;
	    break;
    }
}

/*
 * Add pseudoheaders that will pass the convert arguments to
 * mhbuild.  They have the form:
 *   MHBUILD_FILE_PSEUDOHEADER-text/calendar: /home/user/Mail/inbox/7
 *   MHBUILD_ARGS_PSEUDOHEADER-text/calendar: reply -accept
 * The ARGS pseudoheader is optional, but we always add it when
 * -convertargs is used.
 */
void
add_convert_header (const char *convert_type, char *convert_arg,
                    char *filename, char *drft) {
    char *field_name;

    field_name = concat (MHBUILD_FILE_PSEUDOHEADER, convert_type, NULL);
    annotate (drft, field_name, filename, 1, 0, -2, 1);
    free (field_name);

    field_name = concat (MHBUILD_ARGS_PSEUDOHEADER, convert_type, NULL);
    annotate (drft, field_name, convert_arg, 1, 0, -2, 1);
    free (field_name);
}
